subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/AbstractModuleDependency.java

```diff
@@ -241,22 +252,62 @@ public ModuleDependency capabilities(Action<? super ModuleDependencyCapabilities
             warnAboutInternalApiUse("capabilities");
             return this;
         }
+        initCapabilities();
+        configureAction.execute(moduleDependencyCapabilities);
+        return this;
+    }
+
+    private void initCapabilities() {
         if (moduleDependencyCapabilities == null) {
             moduleDependencyCapabilities = objectFactory.newInstance(
                 DefaultMutableModuleDependencyCapabilitiesHandler.class,
                 capabilityNotationParser
             );
         }
-        configureAction.execute(moduleDependencyCapabilities);
-        return this;
     }
 
     @Override
     public List<Capability> getRequestedCapabilities() {
```

Comment:
Looks like this change in behavior breaks the Palantir consistent versions plugin: 

https://github.com/palantir/gradle-consistent-versions/blob/3a0e254015e020fdacc05e798a24a690355a2fad/src/main/java/com/palantir/gradle/versions/VersionsLockPlugin.java#L350

https://builds.gradle.org/buildConfiguration/Gradle_Master_Check_SmokeTest_SmokeTestsMaxJavaVersion_virtual_Batch_3_1/88356161?expandBuildDeploymentsSection=false&hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildProblemsSection=true&expandBuildTestsSection=true&expandBuildChangesSection=true

Reply:
They expect all capabilities requested on a project dependency to be present in some configuration of the target project (by looking at the configuration container of the other project from the current project) 

There are a couple broken things:
1. CapabilityAdapter does not implement an equals/hashCode. we can fix this
2. The capabilities returned here do not have versions even though the user may include a version when requesting a capability on a project dependency. We forget about that version and it would likely be annoying to remember that version even though we ignore it. But, without remembering that version, we will never satisfy their requirements where the requested capability == declared capability, as declared capabilities must have versions

Reply:
Compatibility was improved here